### ISSUE TITLE
My Issues with debugger

### ISSUE THREAD

**Issue by xNxExOx** [NationalSecurityAgency/ghidra#3049]
**Describe the bug**
I will list multiple problems I have when I am trying to debug `BattleForge.exe` which is x86 windows game.

**To Reproduce**
1. Go to 'Debugger'
2. Click on 'Launch (X)'
3. Enter cmd `F:\Hry\BattleForgeGH\BattleForge.exe -online` and Launch
4. Resume (F5)
5. See error
![image](https://user-images.githubusercontent.com/1893229/118921066-cc7cc180-b937-11eb-81d4-57ce8f2d485a.png)
After I clicked the right button a lot of times (when I was writing this to get an image) it changed to:
![image](https://user-images.githubusercontent.com/1893229/118922676-b9b7bc00-b93a-11eb-9b32-cbd8b17c26dd.png)

**Expected behavior**
No error, or just resolved at first attempt

**Environment (please complete the following information):**
 - OS: Windows 10.0.19042
 - Java Version: openjdk 11.0.9.1 2020-11-04
 - Ghidra Version: 10.0 master (https://github.com/NationalSecurityAgency/ghidra/commit/296360eb3479e70eb734a2abf2eded06b84e3d69)
 - Ghidra Origin: locally built

**Additional context**
Why do I need to click it so many times? Te second one also disappeared after a lot of tries.
After many retries I noticed this:
![image](https://user-images.githubusercontent.com/1893229/118924557-92162300-b93d-11eb-96a2-89976010483e.png)
I think that missing `dbghelp.dll` could be cause of some of these issues, but the file exist.
Is there already any guide on how to debug something bigger than few lines long test app?
Is there any better way of communicating, than github issues?

**Comment by d-millar** [comment]
@xNxExOx So, the module-loading dialog is technically more of a "feature" than a "bug", but I agree it's annoying and we've made changes recently to it, so, with luck, you'll find it less so as well.  It's basically indicating that you are executing within and have not imported that module, in your case ntdll.dll, and is prompting you to do so.  One of those buttons allows you to find and import it;  the other, if I remember correctly, ignores it but only for the current step.  More to the point, in the Modules window (somewhere in the toolbar) is a button to toggle this feature off forever, which I expect you'd prefer.  I would tell you exactly where it is, but, as mentioned, we've changed the behavior and I'm having a little trouble id'ing the old button.

Re communication, github is probably the best.  It will certainly get our attention the fastest.  If you have general question, I'd recommend the Discussions area.  If you mark things as Debugger, we try to get on it as quick as we can.

**Comment by xNxExOx** [comment]
ok so I just misundestood the button description I think it will lik it in sence, that it will use the actual dll and do not bother me with that question again. I solved it by adding `ntdll.dll` to the project, but the app still did not want to start.

After few more tries I figured out the issue, it is breaking on all exceptions.

[runBF.csv](https://github.com/NationalSecurityAgency/ghidra/files/6516414/runBF.csv)
After I put it to file I could find the lines
`"","***ExecutionStatus: BREAK","[]","15:46:37.525"`
`"","***Exception: agent.dbgeng.dbgeng.DebugExceptionRecord64@76fe3313\, first=true","[]","15:46:37.524"`
but it is quite far from the top, so in gui I did not see it, even when I scrolled a bit.

And to my original assumption, what is wrong, well *it did not work*\* and this "error" (red triangle) was on the top of the list, so I assumed it is the reason why it is not working, but actual reason why it does not work, is that it breaks on every exception, and fact it break on exception is not even visible to the user (or at least I did not seen it).

### My question is, how to disable breaking on every single exceptions?
Because if it will be breaking this often the window will never show up :(

**Comment by d-millar** [comment]
So, currently, we do not have a GUI interface to set filters for which exceptions to break on and which to ignore.  You should be able to set these options in the Interpreter window, which accepts pretty much all of the normal windbg commands.  I don't remember off-hand what that command is.  Give me a minute - will check.

**Comment by xNxExOx** [comment]
:( I actually used windbg directly only few times in my life, and IDA make this step about unchecking checkboxes :(
is gui planned?

**Comment by d-millar** [comment]
On the list, but....  The windbg commands are sxe, sxd, sxo, and sxi for enable, disable, output, and ignore.  If I remember correctly, the format is just "sx* #", so "sxd 0xc0000005"  would disable all page-fault handling.  You could also disable it in in the code, but that's probably not a great idea.  The IDA GUI is definitely convenient, but I'd argue there's a lot of functionality in the underlying engine that will probably never make it to either out GUI or IDA's, so you might want to familiarize yourself with some bits.  Or ask here!

**Comment by xNxExOx** [comment]
![image](https://user-images.githubusercontent.com/1893229/118997575-0b873300-b989-11eb-8bca-28a74f16e44c.png)
When I was working with windbg and it break on exception, it at least tell me details about it :( here I see nothing :(
or am I blind and it is somewhere here?

**Comment by d-millar** [comment]
What's in your Interpeter window?

**Comment by d-millar** [comment]
Also, I can't quite see the disasm at RIP - any chance it says 0xCC INT3?

**Comment by xNxExOx** [comment]
```
(kd2)sxd 0xc0000005
ModLoad: 00007ff8`c2570000 00007ff8`c2765000   ntdll.dll
 ModLoad: 00000000`773d0000 00000000`77573000   ntdll.dll
 ModLoad: 00007ff8`c2190000 00007ff8`c21e9000   C:\Windows\System32\wow64.dll
 ModLoad: 00007ff8`c09f0000 00007ff8`c0a73000   C:\Windows\System32\wow64win.dll
 ModLoad: 00000000`773c0000 00000000`773ca000   C:\Windows\System32\wow64cpu.dll
 ModLoad: 00000000`76460000 00000000`76550000   C:\Windows\SysWOW64\KERNEL32.DLL
 ModLoad: 00000000`75750000 00000000`75964000   C:\Windows\SysWOW64\KERNELBASE.dll
 ModLoad: 00000000`751e0000 00000000`7527f000   C:\Windows\SysWOW64\apphelp.dll
 ModLoad: 00000000`76550000 00000000`766e6000   C:\Windows\SysWOW64\USER32.dll
 ModLoad: 00000000`61b40000 00000000`61b68000   F:\Hry\BattleForgeGH\pcre3.dll
 ModLoad: 00000000`75d70000 00000000`75d88000   C:\Windows\SysWOW64\win32u.dll
 ModLoad: 00000000`751b0000 00000000`751d8000   C:\Windows\SysWOW64\WINMM.dll
 ModLoad: 00000000`76390000 00000000`7644f000   C:\Windows\SysWOW64\msvcrt.dll
 ModLoad: 00000000`76190000 00000000`761b3000   C:\Windows\SysWOW64\GDI32.dll
 ModLoad: 00000000`77220000 00000000`772fc000   C:\Windows\SysWOW64\gdi32full.dll
 ModLoad: 00000000`756d0000 00000000`7574b000   C:\Windows\SysWOW64\msvcp_win.dll
 ModLoad: 00000000`75280000 00000000`753a0000   C:\Windows\SysWOW64\ucrtbase.dll
 ModLoad: 00000000`75c00000 00000000`75c7a000   C:\Windows\SysWOW64\ADVAPI32.dll
 ModLoad: 00000000`75fc0000 00000000`76035000   C:\Windows\SysWOW64\sechost.dll
 ModLoad: 00000000`75520000 00000000`755e0000   C:\Windows\SysWOW64\RPCRT4.dll
 ModLoad: 00000000`767b0000 00000000`76d63000   C:\Windows\SysWOW64\SHELL32.dll
 ModLoad: 00000000`76d70000 00000000`76dd3000   C:\Windows\SysWOW64\WS2_32.dll
 ModLoad: 00000000`6b4c0000 00000000`6b53f000   C:\Windows\SysWOW64\DSOUND.dll
 ModLoad: 00000000`74f00000 00000000`74f08000   C:\Windows\SysWOW64\VERSION.dll
 ModLoad: 00000000`6b040000 00000000`6b057000   C:\Windows\SysWOW64\USP10.dll
 ModLoad: 00000000`75970000 00000000`75bf2000   C:\Windows\SysWOW64\combase.dll
 ModLoad: 00000000`72ac0000 00000000`72b04000   C:\Windows\SysWOW64\powrprof.dll
 ModLoad: 00000000`02480000 00000000`024c4000   C:\Windows\SysWOW64\powrprof.dll
 ModLoad: 00000000`6b900000 00000000`6b91d000   C:\Windows\SysWOW64\winmmbase.dll
 ModLoad: 77300000 77325000   C:\Windows\SysWOW64\IMM32.DLL
 ModLoad: 00000000`72ab0000 00000000`72abd000   C:\Windows\SysWOW64\UMPDC.dll
```
I tried `sxd 0xc0000005` before starting debugging, but it is not enough, I was hoping to see the exception number somewhere, to just add it.

**Comment by xNxExOx** [comment]
![image](https://user-images.githubusercontent.com/1893229/119000016-05925180-b98b-11eb-8b9f-fc21c8fc229f.png)
if I am doing it correctly, then there is an issue :( because the address does not exist.

**Comment by d-millar** [comment]
Right, that dialog is actually a Sleigh interpreter so you'd need to add ":8" to the end.  Probably easier to just scroll to it in the Dynamic window or in the Register window, add the Type "pointer" to RIP and click it.

**Comment by xNxExOx** [comment]
🤔 could it be because of the weird architecture that I did not fully understand, that 32 bit app runs against 64 bit kernel, then the actual address might be 64 bit, but even if I change it to  `0x7ff8c21ab530:8` it still not exist :(
![image](https://user-images.githubusercontent.com/1893229/119001416-3e7ef600-b98c-11eb-9a22-311e63bedd00.png)
I changed it to pointer, but clicking on it does nothing :(

**Comment by d-millar** [comment]
Hmmm, maybe double-click?  Leaving that aside, maybe in the Interpreter window, try "d 0x7ff8c21ab530"

**Comment by xNxExOx** [comment]
I tried single and double click to everything in that row, except the checkbox :(
```
 (kd2)d 0x7ff8c21ab530
00007ff8`c21ab530  c6 44 24 20 01 eb 05 48-8b 5c 24 50 65 48 8b 04  .D$ ...H.\$PeH..
00007ff8`c21ab540  25 30 00 00 00 48 89 98-90 14 00 00 8a 44 24 20  %0...H.......D$ 
00007ff8`c21ab550  e9 eb 66 fe ff cc 83 ea-78 74 26 81 ea a0 84 00  ..f.....xt&.....
00007ff8`c21ab560  00 74 15 81 fa 00 24 00-00 74 04 33 c0 c3 cc 48  .t....$..t.3...H
00007ff8`c21ab570  8d 05 ea eb 01 00 c3 cc-48 8d 05 c1 eb 01 00 c3  ........H.......
00007ff8`c21ab580  cc 48 8d 05 78 ea 01 00-c3 cc 4c 21 74 24 30 4c  .H..x.....L!t$0L
00007ff8`c21ab590  8d 4d 40 4c 21 75 48 48-8d 4d 38 bf 00 40 00 00  .M@L!uHH.M8..@..
00007ff8`c21ab5a0  c7 44 24 28 00 00 00 08-be 04 00 00 00 48 89 7d  .D$(.........H.}
```
🤔 is this expected output?

**Comment by d-millar** [comment]
Yep, that's what you want, although not what I was expecting.  Hold on - one moment....

**Comment by d-millar** [comment]
Try ".exr -1" in the Interpreter window - should list the last exception.

**Comment by xNxExOx** [comment]
```
 (kd2).exr -1
ExceptionAddress: 000000007587a6e2 (KERNELBASE!RaiseException+0x0000000000000062)
   ExceptionCode: 406d1388 (Visual C++ exception)
  ExceptionFlags: 00000000
NumberParameters: 4
   Parameter[0]: 0000000000001000
   Parameter[1]: 0000000001e1976c
   Parameter[2]: 000000000000320c
   Parameter[3]: 0000000000000000
```

**Comment by xNxExOx** [comment]
this really nead's a gui with button "Ignore All Exceptions"

**Comment by d-millar** [comment]
Take your point.  :)  I don't think you need to pay attention to this one, so try "sxi 0x406d1388" or "sxd 0x406d1388" or maybe even "sxi vcpp" (I think that ignores one class of Visual C++ exceptions.

**Comment by xNxExOx** [comment]
```
 (kd2)sxi vcpp
// I clicked "Resume (F5)" here, and it break again
(kd2).exr -1
ExceptionAddress: 000000007587a6e2 (KERNELBASE!RaiseException+0x0000000000000062)
   ExceptionCode: 406d1388 (Visual C++ exception)
  ExceptionFlags: 00000000
NumberParameters: 4
   Parameter[0]: 0000000000001000
   Parameter[1]: 0000000001e1976c
   Parameter[2]: 000000000000320c
   Parameter[3]: 0000000000000000
 (kd2)sxi 0x406d1388
(kd2)sxd 0x406d1388
// I clicked "Resume (F5)" here, and it run a bit
ModLoad: 00000000`739b0000 00000000`73fb9000   C:\Windows\SysWOW64\windows.storage.dll
 ModLoad: 00000000`73980000 00000000`739a4000   C:\Windows\SysWOW64\Wldp.dll
 ModLoad: 00000000`77330000 00000000`773b7000   C:\Windows\SysWOW64\SHCORE.dll
 ModLoad: 00000000`753c0000 00000000`75405000   C:\Windows\SysWOW64\shlwapi.dll
 ModLoad: 00000000`750d0000 00000000`750e8000   C:\Windows\SysWOW64\profapi.dll
 ModLoad: 00000000`01f00000 00000000`01f16000   C:\Windows\SysWOW64\xinput1_3.dll
 ModLoad: 00000000`01f00000 00000000`01f16000   C:\Windows\SysWOW64\xinput1_3.dll
 ModLoad: 00000000`76de0000 00000000`77215000   C:\Windows\SysWOW64\SETUPAPI.dll
 ModLoad: 00000000`75690000 00000000`756cb000   C:\Windows\SysWOW64\cfgmgr32.dll
 ModLoad: 00000000`75d90000 00000000`75da9000   C:\Windows\SysWOW64\bcrypt.dll
 ModLoad: 00000000`6a590000 00000000`6a71f000   C:\Windows\SysWOW64\D3D9.dll
 ModLoad: 00000000`73fc0000 00000000`73fcf000   C:\Windows\SysWOW64\kernel.appcore.dll
 ModLoad: 00000000`738d0000 00000000`738f6000   C:\Windows\SysWOW64\dwmapi.dll
 ModLoad: 00000000`6a1c0000 00000000`6a588000   C:\Windows\SysWOW64\D3DX9_38.dll
 ModLoad: 00000000`73900000 00000000`73974000   C:\Windows\SysWOW64\uxtheme.dll
 ModLoad: 00000000`6a0e0000 00000000`6a1b9000   C:\Windows\System32\DriverStore\FileRepository\nv_dispi.inf_amd64_d71d3f5ea7618cbb\nvldumd.dll
 ModLoad: 00000000`75110000 00000000`7511e000   C:\Windows\SysWOW64\msasn1.dll
 ModLoad: 00000000`75120000 00000000`75146000   C:\Windows\SysWOW64\cryptnet.dll
 ModLoad: 00000000`75410000 00000000`75511000   C:\Windows\SysWOW64\CRYPT32.dll
 ModLoad: 00000000`75150000 00000000`7515a000   C:\Windows\SysWOW64\cryptbase.dll
 ModLoad: 00000000`75f70000 00000000`75fb7000   C:\Windows\SysWOW64\WINTRUST.DLL
 ModLoad: 00000000`753a0000 00000000`753b9000   C:\Windows\SysWOW64\imagehlp.dll
 ModLoad: 00000000`75190000 00000000`751a3000   C:\Windows\SysWOW64\CRYPTSP.dll
 ModLoad: 00000000`75160000 00000000`7518f000   C:\Windows\SysWOW64\rsaenh.dll
 ModLoad: 00000000`75db0000 00000000`75e0d000   C:\Windows\SysWOW64\bcryptPrimitives.dll
 ModLoad: 00000000`68c90000 00000000`6a0dd000   C:\Windows\System32\DriverStore\FileRepository\nv_dispi.inf_amd64_d71d3f5ea7618cbb\nvd3dum.dll
 ModLoad: 00000000`68c20000 00000000`68c85000   C:\Windows\SysWOW64\FaultRep.dll
 ModLoad: 00000000`72b40000 00000000`72cc8000   C:\Windows\SysWOW64\dbghelp.dll
 ModLoad: 00000000`68bf0000 00000000`68c16000   C:\Windows\SysWOW64\dbgcore.DLL
 (28cc.338c): Visual C++ exception  - code 406d1388 (first chance)
 First chance exceptions are reported before any exception handling.
 This exception may be expected and handled.
```
but then break again, and it seems like with the same code :(

**Comment by d-millar** [comment]
Hmmm, you might try just sxi, but, honestly, that should have done the trick.  I'm a little perplexed.

If you're really feeling bold, you could also edit the code - DbgManagerImpl::line 607, change:

                statusMap.put(DbgExceptionEvent.class, DebugStatus.BREAK);
to
		statusMap.put(DbgExceptionEvent.class, DebugStatus.NO_CHANGE);

That will ignore all exceptions.

**Comment by xNxExOx** [comment]
```
 (kd2)sxi
Numeric expression missing from '<EOL>'
```
that means edit the code, recompile, and extract the zip again? or is it some "runtime" loading?

**Comment by d-millar** [comment]
Sorry, not quite clear again in my instructions:

sxi 0x406d1388

Rather than:

sxi 0x406d1388
sxd 0x406d1388

which is what I think you entered.

Sxd will override sxi - so the second sequence say "ignore 0x406d1388 - no, change my mind, disable 0x406d1388"

I think you've effectively tried "sxi vcpp" and "sxd 0x406d1388", but not "sxi 0x406d1388".
Does that make sense?

**Comment by d-millar** [comment]
Oh, and right - yeah means "edit, recompile, and extract" unless you're running out of Eclipse in developer mode

**Comment by xNxExOx** [comment]
```
 (kd2)sxi 0x406d1388
(kd2)g
(kd2).exr -1
ExceptionAddress: 000000007587a6e2 (KERNELBASE!RaiseException+0x0000000000000062)
   ExceptionCode: 406d1388 (Visual C++ exception)
  ExceptionFlags: 00000000
NumberParameters: 4
   Parameter[0]: 0000000000001000
   Parameter[1]: 0000000001e19834
   Parameter[2]: 00000000000009a8
   Parameter[3]: 0000000000000000
```
I decided to try `g` so it is in the log, and it seemed to run again for a short amount of time.

**Comment by d-millar** [comment]
Wow, really doesn't want to ignore that does it.  Does ignoring just that exception work in IDA?  (Also, would be curious about whether it worked in windbg, but I realize that's not your thing.)

**Comment by xNxExOx** [comment]
🤔 this is weird.
WinDBG GUI work in tersm, that when I do `sxi` it does not report that exception any more, but 
```
0:000> sxi 0xc0000005
0:000> sxi 0xe06d7363
0:000> g
ModLoad: 77300000 77325000   C:\Windows\SysWOW64\IMM32.DLL
ModLoad: 72ab0000 72abd000   C:\Windows\SysWOW64\UMPDC.dll
ModLoad: 739b0000 73fb9000   C:\Windows\SysWOW64\windows.storage.dll
ModLoad: 73980000 739a4000   C:\Windows\SysWOW64\Wldp.dll
ModLoad: 77330000 773b7000   C:\Windows\SysWOW64\SHCORE.dll
ModLoad: 753c0000 75405000   C:\Windows\SysWOW64\shlwapi.dll
ModLoad: 750d0000 750e8000   C:\Windows\SysWOW64\profapi.dll
ModLoad: 020e0000 020f6000   C:\Windows\SysWOW64\xinput1_3.dll
ModLoad: 020e0000 020f6000   C:\Windows\SysWOW64\xinput1_3.dll
ModLoad: 76de0000 77215000   C:\Windows\SysWOW64\SETUPAPI.dll
ModLoad: 75690000 756cb000   C:\Windows\SysWOW64\cfgmgr32.dll
ModLoad: 75d90000 75da9000   C:\Windows\SysWOW64\bcrypt.dll
ModLoad: 689f0000 68b7f000   C:\Windows\SysWOW64\D3D9.dll
ModLoad: 73fc0000 73fcf000   C:\Windows\SysWOW64\kernel.appcore.dll
ModLoad: 738d0000 738f6000   C:\Windows\SysWOW64\dwmapi.dll
ModLoad: 67bf0000 67fb8000   C:\Windows\SysWOW64\D3DX9_38.dll
ModLoad: 73900000 73974000   C:\Windows\SysWOW64\uxtheme.dll
ModLoad: 68910000 689e9000   C:\Windows\System32\DriverStore\FileRepository\nv_dispi.inf_amd64_d71d3f5ea7618cbb\nvldumd.dll
ModLoad: 75110000 7511e000   C:\Windows\SysWOW64\msasn1.dll
ModLoad: 75120000 75146000   C:\Windows\SysWOW64\cryptnet.dll
ModLoad: 75410000 75511000   C:\Windows\SysWOW64\CRYPT32.dll
ModLoad: 75150000 7515a000   C:\Windows\SysWOW64\cryptbase.dll
ModLoad: 75f70000 75fb7000   C:\Windows\SysWOW64\WINTRUST.DLL
ModLoad: 753a0000 753b9000   C:\Windows\SysWOW64\imagehlp.dll
ModLoad: 75190000 751a3000   C:\Windows\SysWOW64\CRYPTSP.dll
ModLoad: 75160000 7518f000   C:\Windows\SysWOW64\rsaenh.dll
ModLoad: 75db0000 75e0d000   C:\Windows\SysWOW64\bcryptPrimitives.dll
ModLoad: 667a0000 67bed000   C:\Windows\System32\DriverStore\FileRepository\nv_dispi.inf_amd64_d71d3f5ea7618cbb\nvd3dum.dll
ModLoad: 688a0000 68905000   C:\Windows\SysWOW64\FaultRep.dll
ModLoad: 72b40000 72cc8000   C:\Windows\SysWOW64\dbghelp.dll
ModLoad: 68870000 68896000   C:\Windows\SysWOW64\dbgcore.DLL
ModLoad: 75e10000 75ee4000   C:\Windows\SysWOW64\MSCTF.dll
ModLoad: 755f0000 75686000   C:\Windows\SysWOW64\OLEAUT32.dll
ModLoad: 73460000 73523000   C:\Windows\SysWOW64\dxgi.dll
ModLoad: 68860000 6886f000   C:\Windows\SysWOW64\resourcepolicyclient.dll
WARNING: Continuing a non-continuable exception
```
and then nothing, it is just stuck there :(
Not really using it, does not mean, I do not know how to start it, if I need to. I needed it once to use time travel debugging on this exact exe, so I know it was working before.
Ok I tried to disable exceptions in GUI instead and it starts without any breaking. 🤔 so GUI is better 😮 or we just do not know how to set it up correctly from console :(

**Comment by xNxExOx** [comment]
I build ghidra with suggested code change, but extracting it takes quite long :( what is the reasoning behind creating a zip file instead of folder? Windows says it will take 20 minutes to extract :(

**Comment by d-millar** [comment]
Interesting, might be worth trying to execute "sxd 0x406d1388" as soon as you get the Interpreter window, i.e. before anything else to see if this has an affect.  Just trying to re-create the scenario that worked for you.

Re zip file, not sure how you are running normally - if your set-up is using the zip, you probably should recreate the zip

**Comment by xNxExOx** [comment]
I have the modified code, and I start with:
```
sxi 0xe06d7363
sxd 0x406d1388
sxi 0xc0000005
```
and it still prints them, but memory usage is bit crazy 😮 and still rising
![image](https://user-images.githubusercontent.com/1893229/119012983-d255bf80-b996-11eb-97ca-68d753f4a9e5.png)

and now it is totally frozen, but new window popped up with and error:
![image](https://user-images.githubusercontent.com/1893229/119013170-07621200-b997-11eb-8ab0-dca103db50ce.png)
Unfortunately that "Details" button does nothing :(

**Comment by xNxExOx** [comment]
oh after quite some time it expanded:
![image](https://user-images.githubusercontent.com/1893229/119013482-50b26180-b997-11eb-8631-490c4690f81b.png)
The memory usage is so weird

**Comment by d-millar** [comment]
I think you slipped back to the Console columns out-of-order bug - is Time last?

**Comment by xNxExOx** [comment]
![image](https://user-images.githubusercontent.com/1893229/119014204-04b3ec80-b998-11eb-90ed-38c6fa88497e.png)
yes it is last, I did not reordered them, I think it have something to do with the insane memory usage, that is over 16 GB now.

**Comment by xNxExOx** [comment]
will not help, I am going to write a new bug about that

**Comment by xNxExOx** [comment]
https://github.com/NationalSecurityAgency/ghidra/issues/3053 here it is.
So I can not close it, and I already tried not having it as active window, does not help either :( it reaches 2 GB memory usage almost instantly, and is more ore less unusable :( because it reacts on clicks with many seconds delays, and slowing down.
No clue what I can try next. I want to give ghidra a try, because of the significant price difference

**Comment by xNxExOx** [comment]
I created a very simple application that throws these exceptions, and every thrown exc eption increases memory usage, by a small ammount. But with 90 lines of code and `EXCEPTION_ACCESS_VIOLATION`, `std::exception`, `std::runtime_error`, and `char*` thrown every second it increases by about 50 MB per 50 s, because of garbage collection it is bit hard to count very accurately, but I assume that game that have 26 MB binary and throws more exceptions, will grow faster :(

**Comment by d-millar** [comment]
Well, you can turn off the tracing to conserve memory, but only the Objects & Interpreter views will be active.

**Comment by xNxExOx** [comment]
I am trying to debug:
```cpp
#include <iostream>
#include <thread>
int main()
{
    std::cout << "Hello World!" << std::endl;
    auto i = 1;
    while (true)
    {
        std::this_thread::sleep_for(std::chrono::seconds(1));
        std::cout << i++ << std::endl;
    }
}
```
But there are two debugger options that actually start it with a console window. (`IN VM MS dbgeng local debugger`, and `IN VM MS dbgmodel local debugger`)
but when I set up a breakpoint on:
`        00414709 83  c1  01       ADD        ECX ,0x1`
`Address`: `00414709`
`Length`: `1`
`Kind`: `SW_Execute`
they behave differently:
`dbgeng`: does not break at all
`dbgmodel`: breaks on `Access violation`
and the memory usage also rising, over 5 GB in just a few minutes. So that leads me to a question where I can disable tracing? I searched options, F1 help, but can not find any mention about it :(
I am really trying to give ghidra a chance, because of the huge price difference compared to IDA, but so far it can not do even simple tasks :(  do you consider it usable? (and/or usable on windows, if the answer should differ)

**Comment by d-millar** [comment]
@xNxExOx Thanks for the sample - will compile it up and see what we can do.  The short answer by the way is no - you are playing with "alpha" code.  It has not been released, so you are operating on the bleeding, as-is, no guarantees edge.  Sorry about that, but, perhaps, take solace in the thought you are making it better!

**Comment by xNxExOx** [comment]
:) I know it is not in a released version. I was just hoping it might be something wrong in my setup that would be fixable :(

**Comment by d-millar** [comment]
Looking at it now - give me a few minutes.  :)

**Comment by d-millar** [comment]
OK, so good news / bad news!

I compiled your code in Visual Studio 2019, ran it from Ghidra, first using dbgeng and then dbgmodel.  I should say for starters one difference in our set-ups is I'm running out of Eclipse, so that I can debug any possible problems and I built your application as Debug for the same reason.

Probably as a result, my breakpoint is at 0x414862 82 C1 01  ADD ECX, 0x1, a half dozen lines or so before the cout call (two instructions before the "endl" push).  My breakpoint hits in both versions on a regular basis.   Also, I am not seeing ANY exceptions thrown (!).

The target process itself was using somewhere between 300-500k.  Ghidra was using 2.4G for dbgeng and 2.8G for dbgmodel, both of which are about what I expect.  Ghidra, as you have noticed , is just a pretty expensive memory user.  That said, after 10 minutes of execution, it was stable at that size.

OK, so what does that all mean.  Well (a) it doesn't help you at all.  :(  It also doesn't really tell me what the cost of the tracing is, because effectively I'm only stopping for the breakpoint events.  FYI, to run without tracing, untoggle "Record Automatically" in the corner pull-down in the Objects window.

Happy to debug this further.  Want to send me your compile settings for the code?

**Comment by xNxExOx** [comment]
here is my binary [HelloWorldTest.zip](https://github.com/NationalSecurityAgency/ghidra/files/6523477/HelloWorldTest.zip)
I will consider installing eclipse and reading how to build it ghidra with it.

**Comment by d-millar** [comment]
Hmmm, still happy as a clam and smaller (1.6G), which suggests one of the following is true:
(1) Running out of Eclipse matters (shouldn't and seems very unlikely)
(2) My version of the debugger is ahead of yours and has some useful fixes (quite possible)
(3) We're doing something different

In case it's the last, this is what my test looks like:
(0) Import and analyze HelloWorldTest.exe
(1) Start the IN-VM dbgeng or dbgmodel agent
(2) Quick launch HelloWorldTest
(3) In the static listing, go to 0x41954f
(4) Hit K to add a SW_EXECUTE bpt
(5) Select the process in the Objects tree and hit Resume
(6) Repeat (5)

**Comment by xNxExOx** [comment]
![image](https://user-images.githubusercontent.com/1893229/119163550-cbdd4b80-ba5b-11eb-8765-13cac9f242a8.png)
I see a difference, I can not do a "Quick launch".
I was always using Launch, because that button is available.

Will there be any possibility to try newer version? Or is it private only? Because public debugger branch is still only behind the master.

**Comment by d-millar** [comment]
What you're doing should work, but for the sake of debugging:
(1) from the FrontEnd (where you have 1&2), drag HelloWorldTest onto the debugger, so that it loads in the Debugger's static Listing window
(2) now quick launch should be enable

**Comment by d-millar** [comment]
Re branches, actually looks like master is ahead of the public debugger branch now (and actually ahead of my repo, although I think identical from a debugger standpoint), so am leaning toward "that's not the problem"

**Comment by xNxExOx** [comment]
😮 it works.
Should I create a new bug with text "it is somehow broken if I do not drag exe to debugger"? That is so weird, for you it works?
I needed to create new project and analyze it again, and I noticed it is cheating a bit if I do it this way it finds PDB, but when I did it by clicking on the icons it did not find it.

**Comment by d-millar** [comment]
Wow, OK - well honestly I wasn't expecting that!  That is so weird!

Alright, I'll play around a it more when I get a chance.  Having the program in the listing should make things a lot easier, but not having it there shouldn't break things.  Your PDB guess is interesting, but keep in mind I don't have your PDB, so...

Really appreciate your toughing it out, by the way!

**Comment by xNxExOx** [comment]
well the code is so simple so I did not care and before it was saying something like it will not load it from the path, because it does not trust it.
I retested it without dragging, and now it works, so the issue is that something was wrong with the analysis? 🤔 
I will continue testing this way, how far I will get

**Comment by d-millar** [comment]
OK, thanks!  I will do the same after I clean up some other stuff.  Technically, the analysis shouldn't matter either (I debug executable I haven't analyzed all the time), so...your guess is as good as mine, maybe better.   :)

**Comment by nsadeveloper789** [comment]
OK, so coming late to this party. I'll try to clarity a few things, and express our design decisions to see how they match with your opinions.

re/ GUI options for toggling breaks on exceptions: We try to avoid platform-specific options in the GUI. Instead we try to have the connector describe its options and find a way to expose those generically in the UI. At the moment, we're really not there. We have a dialog for generic method invocation, but none of the connectors support it yet. We've also briefly explored configuring an already-running model. At the moment, connection time is the only moment you can tweak options. Once running, the only option we currently have is decimal vs hexadecimal display of numbers. In any case, it is an area of exploration, so a bank of checkboxes for exceptions to ignore may be a ways off. We intended (at least in the short term) for most platform-specific configuration, tweaks, actions, etc., to be performed via the interpreter.

re/ RIP not being in memory: I've seen this happen once when RIP was inside dbghelp.dll, which is hidden by dbgeng. If this comes up again, try running `!address @RIP` in the interpreter. We don't have a built-in way around this, yet, but it is possible to write a script that forces the hidden region into Ghidra's memory map. We're considering a button that lets Ghidra ignore the memory map, but toggling it could lead to a bunch of failed memory reads, since it would allow the view to navigate outside of mapped memory.

re/ The console's memory use: The Debug Console should have a maximum buffer of 100 items. Either that limit is not implemented correctly, it's been modified from its default, or something else is eatign up the memory. Certainly, the crash from closing it is a bug. Tracing caches in memory, but ultimately, writes all the data to disk, so it shouldn't be exploding the memory footprint. That said, I haven't done much testing with an eye on the memory footprint.

I think I missed the punchline in all of this. Dragging the program onto the "Debugger" tool is what's causing the issue? vice what, just launching by specifying the path? OK. So the functional difference is using Quick Launch vs. Launch....?

**Comment by nsadeveloper789** [comment]
Yes, the live listing should show live disassembly, even if the code has modified itself. If you set breakpoints in that state, double-check the address in the breakpoints view. If it was set using the static listing, it could suffer the same off-cut problem.

**Comment by d-millar** [comment]
@nsadeveloper789 So I don't think we know the exact punchline yet....the problem appeared to be solved by having the target in the static listing vs. launching it via the command line without it being resident in the tool.  Might be masking the real problem though.  @xNxExOx re @nsadeveloper789's comment, just to clarify - Dynamic Listing should be accurate for true bytes, Static Listing reflects the state of the executable at the time of import (or if modified through the Byte Viewer interface).

**Comment by xNxExOx** [comment]
re/ GUI it would be great for people like me assuming it would remember the values, otherwise I can just paste the list of commands to the interpreter at the start. In my opinion Connector can provide list of `name`:`string`, `description/tooltip`:`string`, `value`:`object` Java should be able to pick GUI element based on type `bool`->checkbox, `int`->numeric field...

re/ RIP not being in memory, yes I sssume it was in `dbghelp.dll` or some other close kernel dll, because it was catching exceptions, and these are triggered for debugger in there. Will try `!address @RIP`

re/ memory use. I was not checking what component is eating the memory, I do not have eclipse installed (yet?). I tried to limit console to 10, so I already verivy, it is not the console gui. My current assumption is that it is related to the Trace recording, which I am going to try next, but first I run analysis on the game again, to eliminate that as a possibility.
Side note for that using WinDBG's feature Time Travel Debugging, which is something similar in functionality (did not check any details yet), I get over 3TB in 45 minutes which is time it takes for window to appear and handle one click.

re/ draging vs selecting. It is just weird, I did not know about the drag option, and I do not see any reason why it should make a difference, but it does. Without dragging it does not see functions, until it executes them, and I start clicking around. If I drag it, I can easily navigate to main, before starting.
My assumption is that something really weird did happen, but no clue how to pinpoint it yet.

re/ listing being accurate only eliminate the improbable possibility, that breakpoints was misplaced, because of "broken" analysis. 
Side note listing being accurate is one of few things, that ghidra have better than IDA, where I can not show live disassembly, I need to refresh it manually and confirm to overwrite the existing one

I am still interested in testing next version, because I hope it will get competitive to IDA, without a price tag. :)

**Comment by xNxExOx** [comment]
so with real game again:
![image](https://user-images.githubusercontent.com/1893229/119172158-dd2b5580-ba65-11eb-9168-2d1cf3ee131d.png)
the gui is frozen again, and memory usage was ~8 GB and after few minutes it rise to 13 GB and then continue to rise very slowly. CPU usage was bit more than 2 cores.

same test with dbgmodel was looking better at about 4.5 GB, until it throw the unhandled exception that looks exactly the same as this one https://github.com/NationalSecurityAgency/ghidra/issues/3053 so my guess for this is, that in the frozen gui it can not find one of the tables? Since the first of these exceptions memory start rasing quickly to ~14 GB at which the speed slow down.

**Comment by xNxExOx** [comment]
oh and with the dbgeng you can see the interpreter window is full of red lines and the scrollbar is small, so I assume quite a lot of lines in there.

**Comment by d-millar** [comment]
OK, so a couple of notes...  The "unable to deliver callback" message gets triggered when you kill one session and restart it without closing Ghidra.  Definitely a bug and on my list.  More interesting is the fact that that only happens when you have a callback, so my working assumption is you're getting one of those for every unhandled exception you hit.  Probably, a better hack in the code would have been to modify the "interest" flags for which exceptions get reported.  I should have thought of this.  I will try to find the windbg command to turn off interest in receiving exceptions.

Separate note:  you mentioned the GUI is frozen.  Is it frozen frozen, or slow, or various functions don't work?  In your capture, it appears that the process is running, so most actions will be disabled.

**Comment by d-millar** [comment]
OK, so you're going to hate me for this, but I would revert your previous change to the code and make the following change:

In DebugEventCallbacks, comment out line 81:
	@ForInterest(DebugEvent.EXCEPTION)

This will cause Exception to NOT be registered via the SetInterestMask call to the engine, i.e. Ghidra will be completely unaware of them and the program will not stop and pass control to the debugger.   Theoretically, this should solve the memory issue.   (Fingers crossed)

**Comment by xNxExOx** [comment]
After maybe half a minute it is slow frozen, like some things change, but user input is processed seconds later.
After more time it is frozen frozen, that gui no longer reacts to anything (at least no in over 60s after clicking), and even the window is no longer scroling messages. And in case of dbgmodel the popup is slow frozen that it takes over a minute for the details to expand.

**Comment by xNxExOx** [comment]
I tried to build debug to test your suggestion :( but I am probably cursed, because it does not work for me https://github.com/NationalSecurityAgency/ghidra/issues/3062 😢

**Comment by xNxExOx** [comment]
I tried `//@ForInterest(DebugEvent.EXCEPTION)` and at first I think it does not work, but it behaves a bit differently.
when I launch it, it does not wait at start, but hit's and stops on it.
When selecting debugger, there is interesting warning in the Debug Console: `
file:/F:/gh/ghidra/Ghidra/Features/Base/bin/main/images/dialog-warning_red.png | start failed | [] | Sat May 22 07:30:43 CEST 2021`
Then after launching interpreter prints:
```
No .natvis files found at C:\Windows\SYSTEM32\Visualizers.
 No .natvis files found at C:\Users\xxxku\AppData\Local\Dbg\Visualizers.
 
Microsoft (R) Windows Debugger Version 10.0.19041.1 AMD64
Copyright (c) Microsoft Corporation. All rights reserved.

 CommandLine: F:\Hry\BattleForgeGH\BattleForge.exe -online
 Unable to add extension DLL:  ntsdexts 
 Unable to add extension DLL:  uext 
 Unable to add extension DLL:  exts 
 SECURE: File not allowed to be loaded -  C:\Windows\SYSTEM32\dbghelp.dll 
Error code:  Win32 error 0n5 
 The call to LoadLibrary( ext ) failed,  Win32 error 0n2 
    " Systém nemůže nalézt uvedený soubor. "
Please check your debugger configuration and/or network access.
 Unable to add extension DLL:  wow64exts 
 
************* Path validation summary **************
 Response                         Time (ms)     Location
 Deferred                                         srv* 
 Symbol search path is:  srv*
 Executable search path is: 
 ModLoad: 00000000`00400000 00000000`01e72000   BattleForge_retail.exe
 ModLoad: 00007ffd`c4e50000 00007ffd`c5045000   ntdll.dll
 ModLoad: 00000000`775c0000 00000000`77763000   ntdll.dll
 ModLoad: 00007ffd`c2eb0000 00007ffd`c2f09000   C:\Windows\System32\wow64.dll
 ModLoad: 00007ffd`c47f0000 00007ffd`c4873000   C:\Windows\System32\wow64win.dll
 ModLoad: 00000000`775b0000 00000000`775ba000   C:\Windows\System32\wow64cpu.dll
 ModLoad: 00000000`770b0000 00000000`771a0000   C:\Windows\SysWOW64\KERNEL32.DLL
 ModLoad: 00000000`76450000 00000000`76664000   C:\Windows\SysWOW64\KERNELBASE.dll
 ModLoad: 00000000`753d0000 00000000`7546f000   C:\Windows\SysWOW64\apphelp.dll
 ModLoad: 00000000`76ad0000 00000000`76c66000   C:\Windows\SysWOW64\USER32.dll
 ModLoad: 00000000`61b40000 00000000`61b68000   F:\Hry\BattleForgeGH\pcre3.dll
 ModLoad: 00000000`76430000 00000000`76448000   C:\Windows\SysWOW64\win32u.dll
 ModLoad: 00000000`753a0000 00000000`753c8000   C:\Windows\SysWOW64\WINMM.dll
 ModLoad: 00000000`77390000 00000000`7744f000   C:\Windows\SysWOW64\msvcrt.dll
 ModLoad: 00000000`75e60000 00000000`75e83000   C:\Windows\SysWOW64\GDI32.dll
 ModLoad: 00000000`771a0000 00000000`7727c000   C:\Windows\SysWOW64\gdi32full.dll
 ModLoad: 00000000`763b0000 00000000`7642b000   C:\Windows\SysWOW64\msvcp_win.dll
 ModLoad: 00000000`76740000 00000000`76860000   C:\Windows\SysWOW64\ucrtbase.dll
 ModLoad: 00000000`75f40000 00000000`75fba000   C:\Windows\SysWOW64\ADVAPI32.dll
 ModLoad: 00000000`768c0000 00000000`76935000   C:\Windows\SysWOW64\sechost.dll
 ModLoad: 00000000`76990000 00000000`76a50000   C:\Windows\SysWOW64\RPCRT4.dll
 ModLoad: 00000000`75520000 00000000`75ad3000   C:\Windows\SysWOW64\SHELL32.dll
 ModLoad: 00000000`75b90000 00000000`75bf3000   C:\Windows\SysWOW64\WS2_32.dll
 ModLoad: 00000000`6b6b0000 00000000`6b72f000   C:\Windows\SysWOW64\DSOUND.dll
 ModLoad: 00000000`74f20000 00000000`74f28000   C:\Windows\SysWOW64\VERSION.dll
 ModLoad: 00000000`69680000 00000000`69697000   C:\Windows\SysWOW64\USP10.dll
 ModLoad: 00000000`76010000 00000000`76292000   C:\Windows\SysWOW64\combase.dll
 ModLoad: 00000000`72ce0000 00000000`72d24000   C:\Windows\SysWOW64\powrprof.dll
 ModLoad: 00000000`73640000 00000000`7365d000   C:\Windows\SysWOW64\winmmbase.dll
 ModLoad: 77360000 77385000   C:\Windows\SysWOW64\IMM32.DLL
 ModLoad: 72ca0000 72cad000   C:\Windows\SysWOW64\UMPDC.dll
 ModLoad: 00000000`73c90000 00000000`74299000   C:\Windows\SysWOW64\windows.storage.dll
 ModLoad: 00000000`73b70000 00000000`73b94000   C:\Windows\SysWOW64\Wldp.dll
 ModLoad: 00000000`774d0000 00000000`77557000   C:\Windows\SysWOW64\SHCORE.dll
 ModLoad: 00000000`75fc0000 00000000`76005000   C:\Windows\SysWOW64\shlwapi.dll
 ModLoad: 00000000`73230000 00000000`73248000   C:\Windows\SysWOW64\profapi.dll
 ModLoad: 00000000`02490000 00000000`024a6000   C:\Windows\SysWOW64\xinput1_3.dll
 ModLoad: 00000000`02490000 00000000`024a6000   C:\Windows\SysWOW64\xinput1_3.dll
 ModLoad: 00000000`76c70000 00000000`770a5000   C:\Windows\SysWOW64\SETUPAPI.dll
 ModLoad: 00000000`76950000 00000000`7698b000   C:\Windows\SysWOW64\cfgmgr32.dll
 ModLoad: 00000000`77450000 00000000`77469000   C:\Windows\SysWOW64\bcrypt.dll
 ModLoad: 00000000`689d0000 00000000`68b5f000   C:\Windows\SysWOW64\D3D9.dll
 ModLoad: 00000000`73c80000 00000000`73c8f000   C:\Windows\SysWOW64\kernel.appcore.dll
 ModLoad: 00000000`73660000 00000000`73686000   C:\Windows\SysWOW64\dwmapi.dll
 ModLoad: 00000000`68600000 00000000`689c8000   C:\Windows\SysWOW64\D3DX9_38.dll
 ModLoad: 00000000`73af0000 00000000`73b64000   C:\Windows\SysWOW64\uxtheme.dll
 ModLoad: 00000000`68520000 00000000`685f9000   C:\Windows\System32\DriverStore\FileRepository\nv_dispi.inf_amd64_d71d3f5ea7618cbb\nvldumd.dll
 ModLoad: 00000000`73580000 00000000`7358e000   C:\Windows\SysWOW64\msasn1.dll
 ModLoad: 00000000`74ea0000 00000000`74ec6000   C:\Windows\SysWOW64\cryptnet.dll
 ModLoad: 00000000`762a0000 00000000`763a1000   C:\Windows\SysWOW64\CRYPT32.dll
 ModLoad: 00000000`73590000 00000000`7359a000   C:\Windows\SysWOW64\cryptbase.dll
 ModLoad: 00000000`77560000 00000000`775a7000   C:\Windows\SysWOW64\WINTRUST.DLL
 ModLoad: 00000000`75e90000 00000000`75ea9000   C:\Windows\SysWOW64\imagehlp.dll
 ModLoad: 00000000`74f00000 00000000`74f13000   C:\Windows\SysWOW64\CRYPTSP.dll
 ModLoad: 00000000`74ed0000 00000000`74eff000   C:\Windows\SysWOW64\rsaenh.dll
 ModLoad: 00000000`77470000 00000000`774cd000   C:\Windows\SysWOW64\bcryptPrimitives.dll
 ModLoad: 00000000`670d0000 00000000`6851d000   C:\Windows\System32\DriverStore\FileRepository\nv_dispi.inf_amd64_d71d3f5ea7618cbb\nvd3dum.dll
 ModLoad: 00000000`67060000 00000000`670c5000   C:\Windows\SysWOW64\FaultRep.dll
 ModLoad: 00000000`733f0000 00000000`73578000   C:\Windows\SysWOW64\dbghelp.dll
 ModLoad: 00000000`67030000 00000000`67056000   C:\Windows\SysWOW64\dbgcore.DLL
 ModLoad: 00000000`77280000 00000000`77354000   C:\Windows\SysWOW64\MSCTF.dll
 ModLoad: 00000000`75ae0000 00000000`75b76000   C:\Windows\SysWOW64\OLEAUT32.dll
 ModLoad: 00000000`72630000 00000000`726f3000   C:\Windows\SysWOW64\dxgi.dll
 ModLoad: 00000000`69b80000 00000000`69b8f000   C:\Windows\SysWOW64\resourcepolicyclient.dll
 (500.3d9c): C++ EH exception  - code e06d7363 (first chance)
 (500.3d9c): C++ EH exception  - code e06d7363 (!!! second chance !!!)
```
And if I do `Resume (F5)` it print few more lines:
```
 WARNING: Continuing a non-continuable exception
 (500.2234): Access violation  - code c0000005 (first chance)
 First chance exceptions are reported before any exception handling.
 This exception may be expected and handled.
```

(My assumption that using eclipse would make rebuilding and running easier, seems wrong, because from what I read, I still need to build it from console, and for some reason it did not build the GhidraDev plugin for eclipse)

Can you please contact me when you will think it is more usable, please?

**Comment by xNxExOx** [comment]
[debugging attempt - attach.csv](https://github.com/NationalSecurityAgency/ghidra/files/7081561/debugging.attempt.-.attach.csv)
[debugging attempt - launch.csv](https://github.com/NationalSecurityAgency/ghidra/files/7081562/debugging.attempt.-.launch.csv)
I tried latest version but still same issues 😢

**Comment by d-millar** [comment]
@xNxExOx sorry for the delayed response!  I don't see anything in the logs you include that indicates an actual problem.  Are you seeing the same behavior running under windbg?  To sidestep the e06d7363 exception (assuming that's what you really want to do), you'll still need to use the windbg commands from the Interpreter window.

**Comment by xNxExOx** [comment]
Yes I used windbg commands, but nothing improved since I last tried. :( 
It was just my IDA support period ended, and I seen there is new version, so I hoped maybe Ghidra can now be compared to IDA, but debugger is still unusable (at least for this game) :(

**Comment by d-millar** [comment]
So just to be clear, using just windbg (no Ghidr), you hit the same e06d7363 exception, marked it as ignore (?), and then page-faulted when you continued?

**Comment by xNxExOx** [comment]
using just windbg I do not encounter any exceptions (that early), the game normally starts without any exception, and I can see the window and interact with the game, without any interaction with the windbg needed (unless I set breakpoints of course). So the issue is specific to ghidra, I again tried both debugger options, but the result seemed the same.

**Comment by d-millar** [comment]
OK, gotcha - I misunderstood.  Any chance I can download this game from somewhere and attempt to re-create your issues?

**Comment by xNxExOx** [comment]
sure https://forum.skylords.eu/index.php?/topic/7149-skylords-reborn-download/

**Comment by d-millar** [comment]
thanks - will try to see if I can debug this later this aft!

**Comment by d-millar** [comment]
May need a pointer here:  your install looks different than mine.  You have BattleForge.exe; I seem to have SkylordsRebornUpdater.exe.  Does the installer drop that somewhere else?

**Comment by xNxExOx** [comment]
the first download button points to [google drive BattleForge.7z](https://drive.google.com/u/0/uc?id=1W3W0WTCkpmnJn43afTwzKnoHukRs59D0&export=download), unless you want to actually play the game client might be enough to reproduce the issue (will try that in a bit)

**Comment by d-millar** [comment]
OK, thanks - that helps.  Am stuck on something else for the moment, but will see if I can get further this evening.

**Comment by d-millar** [comment]
OK, so am running that executable in Windbg Preview, and it behaves exactly as it does in Ghidra, i.e. I hit the initial breakpoint (0x80000003), I hit the C++ EH exception (0xe06d7363), and then hits the page fault (0xc0000005), and dies.  Could you check your exception settings for the C++ exception in windbg?  Under Settings->Events & exceptions->Exceptions->C++ EH, I have Break Status == Second Chance Break and Handling Status == Handled.  Also, what do you have for Access Violation?

**Comment by xNxExOx** [comment]
- Access Violation: Break, Not Handled
- C++ EH exception: Second Chance Break, Not Handled
- Visual C++ exception: Ignore, Handled
"Reset to default" button is greyed out, so this might be the default values?

I see it like this, really long time ago I was able to "configure" Windbg and never needed to touch the configuration again (if it was not good by default, really do not remember). 

I remember IDA thou I needed to touch the configuration **once**, do not remember what did I set, but I did not need to change the configuration ever since.

But with ghidra I am not able to disable exceptions not even for single run, no matter what commands I use it always breaks.

**Comment by d-millar** [comment]
Yep, I get it - and am definitely interested in fixing this for you.  I just need to know what the actual problem is to fix it.  Differing behavior between Ghidra and Windbg takes me down one path;  failure to read/save configuration takes me down another.  Unfortunately, our exception settings seem to be identical, so that's probably not it.  The page fault I'm getting is in CSslConnectionImplOpenSsl::SetTrustedRootCertificate at 0x015fba1a, which suggests a completely different problem.  I use "-online" means I'm connecting to a server somewhere.  Didn't read the webpage instructions in great detail - do you know if I have to do something to set up authentication?

**Comment by xNxExOx** [comment]
without the "-online" it tries to start "bootstrapper.exe" or something like that that was EA's updater, and when updater checked game is up to date it started "battlegorge.exe -online" to start the game, it is EA's way to make sure that no matter how user starts the game it will be up to date.
`0x015fba1a` is in `memcopy`, not `CSslConnectionImplOpenSsl::SetTrustedRootCertificate` which start at `.text:012D64B0` and ends `.text:012DC841`, unfortunately Windbg do not have symbols (or decompiled data) so it shows everything as the closest exported function and some offset :(

**Comment by d-millar** [comment]
So, it looks like the issue is the start directory, which in windbg you're probably setting in the initial File->Launch window.  Looking into how to set that from the command line for Ghidra now.

**Comment by d-millar** [comment]
Yeah, so there are a number of possible solutions here, but I haven't found one that actually works yet.  We don't invoke windbg directly, but there may be an API call that allows you to pass parameters.  If so, we could enable the equivalent of "-c" for windbg.  I would have thought .srcpath or .lsrcpath would also solve the problem, but I haven't had any luck on that front yet.  There may also be some equivalent of .gdbinit for windbg; I can't remember whether windbg will look for something like that by default, however, or whether it needs to be passed on the command line.  In any case, our exception reporting story is still crap.  When I run in the failing cases, the exception isn't even reported to the Interpreter window, which seems bad.  I think we already have an open ticket for this, but I am happy to open another one.  One problem that unfortunately is slowing up a solution for you - your game is a good test case, but I can't run it in my normal work environment.  So, again, apologies for the delay on this!

**Comment by xNxExOx** [comment]
No problem :) glad to help (at least by providing a test case) to make Ghidra at least comparable to IDA :)

**Comment by d-millar** [comment]
A little more info:  .createdir does the right thing.  If you start up dbgeng, start BattleForge, and in the Interpreter issue the command:
      .createdir F:\Hry\BattleForgeGH
the game will start (eventually).  The still-very-annoying bit, it seems to be stopping on ThreadCreate events for reasons I don't yet understand, so you have to hit resume two dozen plus times.  I am still looking into this.  

I have also put in tickets for better event/exception views and controls and for possible start-up script processing, so more on that as things get fixed.

**Comment by milandamen** [comment]
I have this same problem. Even setting windbg to ignore all exceptions (see: https://stackoverflow.com/questions/28306310/windbg-how-to-set-all-exception-to-be-passed-into-app ) still causes Ghidra to pause on vcpp exceptions.

**Comment by milandamen** [comment]
@d-millar I am using Ghidra Version 10.1-BETA Build PUBLIC 2021-Nov-16 1948 EST with Java Version 17.0.1.

I am debugging a game called Trackmania 2020 but it is somewhat difficult to set up.

I used the following command to try to suppress exceptions:

```
.foreach(exc {.echo "ct et cpr epr ld ud ser ibp iml out av asrt aph bpe bpec eh clr clrn cce cc dm dbce gp ii ip dz iov ch hc lsq isc 3c svh sse ssec sbo sov vs vcpp wkd rto rtt wob wos *"}) {.catch{sxi ${exc}}}
```

Which results in the following rules:

```
0:000>sx
  ct - Create thread - ignore
  et - Exit thread - ignore
 cpr - Create process - ignore
 epr - Exit process - ignore
  ld - Load module - ignore
  ud - Unload module - ignore
 ser - System error - ignore
 ibp - Initial breakpoint - ignore
 iml - Initial module load - ignore
 out - Debuggee output - ignore
 
  av - Access violation - ignore - not handled
asrt - Assertion failure - ignore - not handled
 aph - Application hang - ignore - not handled
 bpe - Break instruction exception - ignore
bpec - Break instruction exception continue - not handled
  eh - C++ EH exception - ignore - not handled
 clr - CLR exception - ignore - not handled
clrn - CLR notification exception - ignore - handled
 cce - Control-Break exception - ignore
  cc - Control-Break exception continue - not handled
 cce - Control-C exception - ignore
  cc - Control-C exception continue - not handled
  dm - Data misaligned - ignore - not handled
dbce - Debugger command exception - ignore - handled
  gp - Guard page violation - ignore  - not handled
  ii - Illegal instruction - ignore - not handled
  ip - In-page I/O error - ignore - not handled
  dz - Integer divide-by-zero - ignore - not handled
 iov - Integer overflow - ignore - not handled
  ch - Invalid handle - ignore
  hc - Invalid handle continue - not handled
 lsq - Invalid lock sequence - ignore - not handled
 isc - Invalid system call - ignore - not handled
  3c - Port disconnected - ignore - not handled
 svh - Service hang - ignore - not handled
 sse - Single step exception - ignore
ssec - Single step exception continue - not handled
 sbo - Security check failure or stack buffer overrun - ignore - not handled
 sov - Stack overflow - ignore - not handled
  vs - Verifier stop - ignore - not handled
vcpp - Visual C++ exception - ignore - handled
 wkd - Wake debugger - ignore - not handled
 rto - Windows Runtime Originate Error - ignore - not handled
 rtt - Windows Runtime Transform Error - ignore - not handled
 wob - WOW64 breakpoint - ignore - handled
  wos - WOW64 single step exception - ignore - handled
 
   * - Other exception - ignore - not handled
```

Here is a video of me reproducing the problem: https://www.youtube.com/watch?v=s4JbMmdL5M0

**Comment by d-millar** [comment]
Super helpful - thanks!  Looking at it now.

**Comment by d-millar** [comment]
Hmmm, interesting - looks like sx is not handled at the dbgeng level, which is very sad.  I will up the priority of this ticket, given there's really no viable workaround.  Thanks for highlighting this!

**Comment by d-millar** [comment]
OK, so we have a solution (pretty sure) and it may make the next release, but, if you're feeling brave/eager and don't mind building from scratch, you can add the fix yourself.  Line 609 in DbgManagerImpl should be changed to:

     		statusMap.put(DbgExceptionEvent.class, DebugStatus.NO_CHANGE);

**Comment by milandamen** [comment]
I can confirm that this works with latest master (bed44bd55dc88c7c6fab5f158cdb4b9daa087519).

Thank you @d-millar and the team!
